### PR TITLE
Ensure integer wrapping instead of preventing overflow/underflow

### PR DESCRIPTION
**Pull request by ChillerDragon**
See also https://github.com/ddnet/ddnet/commit/65cb2ed7de102f155cde64b4e75839d15b6fe35a
And #5646

## Checklist

- [ ] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/engine/shared/snapshot.cpp [hunk 1]
```diff
@@ -243,7 +243,8 @@ void CSnapshotDelta::UndiffItem(const int *pPast, const int *pDiff, int *pOut, i
 		{
 			unsigned char aBuf[CVariableInt::MAX_BYTES_PACKED];
 			unsigned char *pEnd = CVariableInt::Pack(aBuf, *pDiff, sizeof(aBuf));
-			*pDataRate += (int)(pEnd - (unsigned char *)aBuf) * 8;
+			// addition with wrapping by casting to unsigned
+			*pDataRate += (unsigned)(pEnd - (unsigned char *)aBuf) * 8;
```
Comment (heinrich5991) [inline src/engine/shared/snapshot.cpp:247]: This doesn't work. The addition is still an `int` addition.
```suggestion
			*pDataRate = (unsigned)*pDataRate + (unsigned)(int)(pEnd - (unsigned char *)aBuf) * 8;
```

But we'd have to make sure that wrapping is actually the correct behavior first.

**Review by Robyt3 — COMMENTED** [review]
AFAICT, the `m_aSnapshotDataRate` and `m_aSnapshotDataUpdates` values are only used for rendering these values in debug mode. The `m_aSnapshotDataUpdates` value is incremented by one for every unpacked item:

https://github.com/ddnet/ddnet/blob/1ff5d5b10f0c32ea6b0e533745a0db573942c0ad/src/engine/shared/snapshot.cpp#L603

which could also overflow (although unlikely).

The addition in the other branch, in which `UndiffItem` is not called, could also cause an overflow:

https://github.com/ddnet/ddnet/blob/1ff5d5b10f0c32ea6b0e533745a0db573942c0ad/src/engine/shared/snapshot.cpp#L601

I guess the most correct would be to clear both variables if either of them overflows. Otherwise the rate-per-updates (`R/U`) value in the debug HUD would be incorrect if only one value would wrap around to 0.

**Comment by heinrich5991** [comment]
We could probably also make them 64-bit integers instead so that they can't overflow.

**Comment by ChillerDragon** [comment]
> We could probably also make them 64-bit integers instead so that they can't overflow.

It is a big array should we really increase its size? I would rather change it from ``int`` to ``unsigned`` to fit more.

File: src/engine/shared/snapshot.cpp (hunk 1 above)
Comment (ChillerDragon) [reply src/engine/shared/snapshot.cpp]: It is not correct to wrap. But it only affects a debug counter. I do not think it is worth it to allocate more memory for it. So lets keep it at wrapping and just make sure that clang doesn't trip on it.

**Comment by heinrich5991** [comment]
I'd prefer to make the code correct, so we never have to worry about it again. We could either fix the overflow behavior to be correct (like @Robyt3 suggested) or increase the integer size to 64 bit for that.

**Comment by ChillerDragon** [comment]
Okay I went with ``uint64_t`` because I do not like wasting cpu cycles on an additional if statement. Lets rather waste a few more kilo bytes of memory.

**Review by Robyt3 — APPROVED** [review]

**Comment by ChillerDragon** [comment]
I am actively waiting for this to be merged by the way. Whats the ETA or do I have to merge in my downstream?

**Review by def- — APPROVED** [review]

**Comment by ChillerDragon** [comment]
:heart:

**Pull request merged**